### ISSUE TITLE
PyTorch-specific rules

### ISSUE THREAD

**Issue by sbrugman** [astral-sh/ruff#12349]
Introducing lints for pytorch expands `ruff` utility in the AI/machine learning space, in addition to Numpy: #2455 and Spark: #7272. Detecting potential errors during linting has the potential to save costly GPU compute.

Rule suggestions:

TOR001 -> initialize directly on device (.to): requires less memory [1]
TOR002 -> initialize directly with dtype (.bool, …): requires less memory [1]
TOR003 -> prefer torch over numpy (.fft, …) inside `torch.nn.Module` when `backward` is missing: gradient graph does not pass through numpy [2]
TOR004 -> unused attribute layer on `torch.nn.Module`: layer is specified but unused
TOR005 -> consider use of Sequential for layers: improves readibility

[1]: https://stackoverflow.com/a/69538206
[2]: https://discuss.pytorch.org/t/retaining-gradients-after-comverting-to-numpy-then-pytorch/194702/2

**Comment by randolf-scholz** [comment]
Also #9664.

**Comment by njzjz** [comment]
There is a flake8 plugin called TorchFix developed by the PyTorch team: https://github.com/pytorch-labs/torchfix

**Comment by sbrugman** [comment]
> There is a flake8 plugin called TorchFix developed by the PyTorch team: https://github.com/pytorch-labs/torchfix

Thanks for sharing! This would be a great base to start from!

**Comment by NeilGirdhar** [comment]
> TOR008 -> Use array agnostic syntax (see also https://github.com/astral-sh/ruff/issues/8615), e.g. use x.mean() rather than torch.mean(x)

I think this is backwards.  The Array API is "array agnostic".  So, `torch.mean` is better than `x.mean`.  And even better would be:
```python
xp = get_namespace(x)
xp.mean(x)
```

**Comment by muslll** [comment]
+1 for TorchFix. It integrates with flake8, so it should be easy to port the rules. It has been officially released now with the update to pytorch 2.5.

**Comment by randolf-scholz** [comment]
Rule: detect usage of $x*\log(x)$, advise users to use special function `torch.special.xlogy` instead. (This bug even occurs in torch source code!)

**Comment by isaaccorley** [comment]
bump -- would also love to use this with ruff!
